<div style="text-align: center;">
یادآوری بهتر کدها ✨
</div>

***for numpy***

In [15]:
import numpy as np

 <div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #110878ff, #002ca6ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 700px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      Types of <code>np.random</code> Methods
    </h2>
  </div>
</div>


💙 **np.random.rand**

Random numbers between 0 and 1 (uniform distribution)

In [16]:
np.random.rand(3,4)
np.random.rand(3)

array([0.69059312, 0.20523436, 0.50812003])

💙 **np.random.randn**

Random Numbers with mean 0 and variance 1 (standard normal distribution)

In [17]:
np.random.randn(7,4)
np.random.randn(2)

array([-0.11231096, -0.6929508 ])

💙 **np.random.randint**

Random integer between [low, high) (meaning high is not included)

In [18]:
np.random.randint(10)
np.random.randint(1,30)
np.random.randint(1,30,10)
np.random.randint(1,30,(3,5))

array([[ 5, 14,  2,  4,  5],
       [14,  4, 10,  6, 21],
       [21, 29, 19, 15, 10]], dtype=int32)

💙 **np.random.normal(loc=0.0, scale=1.0, size=None)**

loc = mean <br>
scale = standard deviation

In [19]:
np.random.normal(10, 2 , 5)

array([ 9.52127738, 11.76577613, 12.97661365,  9.50270633,  9.23654444])

💙 **np.random.random(size=None)**

Returns random floats in the half-open interval [0.0, 1.0]

In [20]:
print(np.random.random() ,
np.random.random(5),
np.random.random((3, 3)),
np.random.random(5) * 10)

0.500422107139803 [0.33671503 0.66009097 0.71500491 0.05535941 0.11374036] [[0.22589098 0.05948719 0.47553502]
 [0.67800652 0.95137472 0.45183642]
 [0.69654065 0.54682469 0.3166353 ]] [1.19551081 2.32124338 2.51052028 7.28904504 5.3301968 ]


💙 **np.random.uniform(low=0.0, high=1.0, size=None)**

This function is used to generate random numbers from a continuous uniform distribution.
That is, all numbers between low and high are selected with equal probability
"rand is only for (0,1).<br>
uniform is more flexible and allows specifying any desired interval

In [21]:
np.random.uniform()
np.random.uniform(2,7)
np.random.uniform(2,7,10)
np.random.uniform(2,7,(3,3))

array([[4.17786471, 5.66518071, 6.49494709],
       [3.97799143, 2.02814092, 6.51999895],
       [6.26256497, 2.19099161, 2.97959976]])

💙 **np.random.choice(a, size=None, replace=True, p=None)**

Random selection from the elements of an array 

Parameters:

a → array or number of elements (0 to n-1) <br>
size → number of samples <br>
replace → whether to sample with replacement or not <br>
p → array of probabilities for each element (sum must = 1)

In [22]:
np.random.choice([1,2,3,4,5])

arr = np.array([1,2,3,4,5])
np.random.choice(arr)
np.random.choice(arr, 3) # With repetition
np.random.choice(arr, 3, replace=False) # The argument p is a list of probabilities that specifies the probability of selecting each element from the array arr.

array([4, 2, 3])

💙 **np.random.shuffle**

Randomly shuffles the order of the elements

In [23]:
arr = np.arange(10)
np.random.shuffle(arr)
arr 

array([6, 8, 1, 9, 5, 0, 7, 2, 3, 4])

💙 **np.random.permutation**

Unlike np.random.shuffle(), which modifies the array in-place, np.random.permutation() returns a new array

In [24]:
arr = np.arange(10)
p = np.random.permutation(arr)
print("Original:", arr)
print("Permutation:", p)

Original: [0 1 2 3 4 5 6 7 8 9]
Permutation: [7 4 2 1 9 5 8 3 6 0]


🔵 np.random.seed (old) and np.random.default_rng (new)


In [25]:
rng1 = np.random.default_rng(42)   
rng2 = np.random.default_rng(42)    
rng3 = np.random.default_rng(7)     

print("rng1:", rng1.integers(1, 10, 5))
print("rng2:", rng2.integers(1, 10, 5))   
print("rng3:", rng3.integers(1, 10, 5))  


rng1: [1 7 6 4 4]
rng2: [1 7 6 4 4]
rng3: [9 6 7 9 6]


<div style="text-align: center;">
  <img src="png/1.png" style="width: 650px; border: 1px solid #ccc; border-radius: 8px;">
</div>


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #fb2493ff, #f461adff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 700px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      Creating and Initializing Arrays in <code>NumPy</code>
    </h2>
  </div>
</div>


💗 **np.zeros**

Creates an array filled with zeros

In [26]:
np.zeros(10)
np.zeros((8,6))

array([[0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.]])

💗 **np.ones**

Creates an array filled with ones

In [27]:
np.ones(3)
np.ones((4,5))

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

💗 **np.full**

Creates an array filled with a specified value

In [28]:
np.full((3,4),7)

array([[7, 7, 7, 7],
       [7, 7, 7, 7],
       [7, 7, 7, 7]])

💗 **np.eye(N, M=None, k=0, dtype=float)**

Identity matrix

N → number of rows <br>
M → number of columns (optional, default = N) <br>
k → diagonal position (default 0 = main diagonal) <bar>
k = 0 → main diagonal <br>
k > 0 → diagonals above the main diagonal <br>
k < 0 → diagonals below the main diagonal <br>
dtype → data type (e.g., int, float) <br>

In [29]:
np.eye(5)
np.eye(3, 5)
np.eye(4, 4, k=-1)

array([[0., 0., 0., 0.],
       [1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.]])

💗 **np.identity(n, dtype=None)**

n → size of the square matrix (number of rows and columns)<br>
dtype → data type (e.g., int, float)

In [30]:
np.identity(5)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

💗 **np.linspace**

A specific number of values between two numbers

In [31]:
np.linspace(1,10,8)

array([ 1.        ,  2.28571429,  3.57142857,  4.85714286,  6.14285714,
        7.42857143,  8.71428571, 10.        ])

💗 **np.empty**

Creates an array of specified dimensions without initializing it.

This means that the value inside the array is random and can contain anything (depending on memory)

In [32]:
np.empty((3,8))
np.empty(6)

array([0., 0., 0., 0., 0., 0.])

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #7c2e2eff, #710514ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 700px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      Data Type Conversion in <code>NumPy</code>
    </h2>
  </div>
</div>


Specifying the data type during creation
```python
a = np.array([1, 2, 3], dtype=np.float32)
b = np.array([1.5, 2.8, 3.1], dtype=np.int32)


Changing the data type after creation
```python
x = np.array([1.2, 3.4, 5.6])
y = x.astype(int) 

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #4c1d76ff, #5a5174ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 700px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      NumPy Array Tricks Cheat Sheet
    </h2>
  </div>
</div>



This cheat sheet includes commonly used NumPy functions for array manipulation.

```python
import numpy as np

a = np.array([[1, 2, 3],
              [4, 5, 6]])
 

# 1️⃣ reshape
# فارسی: تغییر شکل آرایه بدون تغییر داده‌ها
# English: Changes the shape of an array without modifying the data
b = a.reshape(3, 2)
 

# 2️⃣ flatten
# فارسی: تبدیل آرایه به 1بعدی، همیشه کپی جدید می‌سازد
# English: Flattens the array into 1D and always returns a copy
f = a.flatten()
 

# 3️⃣ ravel
# فارسی: تبدیل آرایه به 1بعدی، view ممکن است بازگردانده شود
# English: Flattens the array into 1D but returns a view if possible
r = a.ravel()
 

# 4️⃣ transpose
# فارسی: جابجایی سطر و ستون‌ها
# English: Swaps rows and columns of the array
t = a.T

# 5️⃣ resize
# فارسی: تغییر شکل آرایه؛ اگر سایز جدید بزرگ‌تر باشد، عناصر تکرار می‌شوند. 
# English: Changes the shape of an array; if the new size is larger, elements are repeated.
a = np.array([1, 2, 3, 4, 5, 6])

r1 = np.resize(a, (2, 3))   # همان تعداد عنصر
r2 = np.resize(a, (3, 4))   # بزرگ‌تر → تکرار عناصر
r3 = np.resize(a, (2, 2))   # کوچک‌تر → برش داده می‌شود

 



### reshape()
- Returns a view when possible  
- `arr.reshape(shape)`  
- Changes affect original data  

### resize()
- Always creates an independent copy  
- `np.resize(arr, shape)`  
- Changes don't affect original data  


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #320947ff, #54305fff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 700px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      Row and Column
    </h2>
  </div>
</div>


<div style="text-align: center;">
  <img src="png/2.jpg" style="width: 650px; border: 1px solid #ccc; border-radius: 8px;">
</div>


<!DOCTYPE html>
<html lang="fa" dir="rtl">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>تفاوت reshape و resize در NumPy</title>
    <style>
        body {
            font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
            line-height: 1.6;
            color: #333;
            background: linear-gradient(135deg, #f5f7fa 0%, #c3cfe2 100%);
            margin: 0;
            padding: 20px;
            display: flex;
            justify-content: center;
            align-items: center;
            min-height: 100vh;
        }
        
        .container {
            max-width: 900px;
            background: white;
            border-radius: 15px;
            box-shadow: 0 10px 30px rgba(0, 0, 0, 0.1);
            overflow: hidden;
        }
        
        header {
            background: #4b6cb7;
            color: white;
            text-align: center;
            padding: 25px 20px;
        }
        
        h1 {
            margin: 0;
            font-size: 28px;
        }
        
        .description {
            padding: 15px 20px;
            background: #e6eeff;
            text-align: center;
            font-size: 18px;
            border-bottom: 2px dashed #4b6cb7;
        }
        
        .content {
            display: flex;
            flex-wrap: wrap;
            padding: 20px;
        }
        
        .card {
            flex: 1;
            min-width: 300px;
            margin: 15px;
            padding: 25px;
            border-radius: 12px;
            box-shadow: 0 5px 15px rgba(0, 0, 0, 0.08);
            transition: transform 0.3s ease;
        }
        
        .card:hover {
            transform: translateY(-5px);
        }
        
        .reshape {
            background: linear-gradient(135deg, #c2e59c 0%, #64b3f4 100%);
            border: 2px solid #3a9bd8;
        }
        
        .resize {
            background: linear-gradient(135deg, #ff9a9e 0%, #fad0c4 100%);
            border: 2px solid #ff7b7f;
        }
        
        h2 {
            color: #2c3e50;
            border-bottom: 2px solid rgba(0, 0, 0, 0.1);
            padding-bottom: 10px;
            text-align: center;
        }
        
        .code {
            font-family: 'Courier New', Courier, monospace;
            background: rgba(255, 255, 255, 0.7);
            padding: 15px;
            border-radius: 8px;
            margin: 15px 0;
            direction: ltr;
            text-align: left;
            overflow-x: auto;
        }
        
        .property {
            background: rgba(255, 255, 255, 0.5);
            padding: 12px;
            border-radius: 8px;
            margin: 10px 0;
            border-right: 4px solid;
        }
        
        .reshape .property {
            border-right-color: #3a9bd8;
        }
        
        .resize .property {
            border-right-color: #ff7b7f;
        }
        
        .footer {
            text-align: center;
            padding: 20px;
            background: #4b6cb7;
            color: white;
            font-size: 18px;
        }
        
        .highlight {
            background: yellow;
            padding: 0 5px;
            border-radius: 3px;
            font-weight: bold;
        }
        
        @media (max-width: 768px) {
            .content {
                flex-direction: column;
            }
        }
    </style>
</head>
<body>
    <div class="container">
        <header>
            <h1>تفاوت reshape و resize در NumPy</h1>
        </header>
        
        <div class="description">
            <p>مقایسه دو تابع مهم برای تغییر شکل آرایه‌ها در کتابخانه NumPy</p>
        </div>
        
        <div class="content">
            <div class="card reshape">
                <h2>تابع reshape</h2>
                <div class="property">
                    <p><strong>ویژگی اصلی:</strong> ایجاد یک <span class="highlight">View</span> از آرایه اصلی</p>
                </div>
                <div class="property">
                    <p><strong>اثر روی داده اصلی:</strong> تغییر در View روی داده اصلی اثر می‌گذارد</p>
                </div>
                <div class="property">
                    <p><strong>حافظه:</strong> از حافظه یکسان استفاده می‌کند (هیچ کپی ایجاد نمی‌شود)</p>
                </div>
                <div class="code">
                    import numpy as np<br>
                    <br>
                    arr = np.array([1, 2, 3, 4, 5, 6])<br>
                    reshaped = arr.reshape(2, 3)<br>
                    <br>
                    print("آرایه اصلی:", arr)<br>
                    print("آرایه تغییر شکل یافته:", reshaped)<br>
                    <br>
                    # تغییر در reshaped روی arr اثر می‌گذارد<br>
                    reshaped[0, 0] = 99<br>
                    print("آرایه اصلی پس از تغییر:", arr)  # خروجی: [99, 2, 3, 4, 5, 6]
                </div>
            </div>
            
            <div class="card resize">
                <h2>تابع resize</h2>
                <div class="property">
                    <p><strong>ویژگی اصلی:</strong> ایجاد یک <span class="highlight">کپی مستقل</span> از آرایه</p>
                </div>
                <div class="property">
                    <p><strong>اثر روی داده اصلی:</strong> تغییر در کپی جدید روی داده اصلی اثر نمی‌گذارد</p>
                </div>
                <div class="property">
                    <p><strong>حافظه:</strong> حافظه جدیدی اختصاص می‌دهد (کپی مستقل ایجاد می‌کند)</p>
                </div>
                <div class="code">
                    import numpy as np<br>
                    <br>
                    arr = np.array([1, 2, 3, 4, 5, 6])<br>
                    resized = np.resize(arr, (2, 3))<br>
                    <br>
                    print("آرایه اصلی:", arr)<br>
                    print("آرایه تغییر اندازه یافته:", resized)<br>
                    <br>
                    # تغییر در resized روی arr اثر نمی‌گذارد<br>
                    resized[0, 0] = 99<br>
                    print("آرایه اصلی پس از تغییر:", arr)  # خروجی: [1, 2, 3, 4, 5, 6]
                </div>
            </div>
        </div>
        
        <div class="footer">
            <p>reshape → گاهی View برمی‌گرداند (روی داده اصلی اثر دارد) | resize → همیشه یک کپی مستقل می‌سازد</p>
        </div>
    </div>
</body>
</html>

In [33]:
a = np.array([[1,2,3,4],
             [5,6,7,8],
             [9,10,11,12],
             [13,14,15,16]])

In [34]:
a[1] # 2nd row

array([5, 6, 7, 8])

In [35]:
a[:,1] # 2nd column

array([ 2,  6, 10, 14])

In [36]:
a[1:3] # 2nd and 3rd row

array([[ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [37]:
a[[1,2,3]] # 2nd, 3rd and 4th row

array([[ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

In [38]:
a[:, 1:3] # 2nd and 3rd column

array([[ 2,  3],
       [ 6,  7],
       [10, 11],
       [14, 15]])

In [39]:
a[:, [2,3]] # 3rd and 4th column

array([[ 3,  4],
       [ 7,  8],
       [11, 12],
       [15, 16]])

In [40]:
a[0::2] # odd rows

array([[ 1,  2,  3,  4],
       [ 9, 10, 11, 12]])

In [41]:
a[1::2] # even rows

array([[ 5,  6,  7,  8],
       [13, 14, 15, 16]])

In [42]:
a[:, 0::2] # odd columns

array([[ 1,  3],
       [ 5,  7],
       [ 9, 11],
       [13, 15]])

In [43]:
a[:, 1::2] # even columns

array([[ 2,  4],
       [ 6,  8],
       [10, 12],
       [14, 16]])

In [44]:
a[2,3] # Choosing a specific house

np.int64(12)

In [45]:
a[1:3, 2:4] # Selecting part of the matrix

array([[ 7,  8],
       [11, 12]])

In [46]:
a[:] # Selecting the whole matrix

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

In [47]:
a[-1] # Last row

array([13, 14, 15, 16])

In [48]:
a[:,-1] # Last column

array([ 4,  8, 12, 16])

In [49]:
a[-2] # Second last row

array([ 9, 10, 11, 12])

In [50]:
a[[1,2] , [1,2]]  

array([ 6, 11])

In [51]:
a = np.arange(20).reshape(5, 4)
row = [1,2,4]
column = [0,2,3]

a[np.ix_(row, column)]

array([[ 4,  6,  7],
       [ 8, 10, 11],
       [16, 18, 19]])

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #320947ff, #54305fff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 100px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.ix_
    </h2>
  </div>
</div>


🌝**np.ix-**

- English: Selects a full **submatrix** using chosen rows × chosen columns  
- فارسی: یک **زیرماتریس کامل** می‌سازد از ترکیب سطرها و ستون‌های انتخابی  

The output is a submatrix, not a flat array. <br>
خروجی زیرماتریس هست، نه آرایه تخت<br>

Can also be used with a condition:<br>
میشه با شرط هم استفاده کرد<br>
```python
rows = np.where(a[:,0] % 8 == 0)[0]
cols = [0,2]
print(a[np.ix_(rows, cols)])

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #de0b6dff, #ef4a97ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 700px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      Main and Secondary Diagonal
    </h2>
  </div>
</div>


In [52]:
import numpy as np
a = np.array([[1, 2, 3],
              [4, 5, 6],
              [7, 8, 9]])

1️⃣ Get Main Diagonal  

Get elements along the main diagonal (top-left → bottom-right)


In [53]:
np.diag(a)

array([1, 5, 9])



Use k to select diagonals above or below main

In [54]:
np.diag(a, k=1) # main diagonal
np.diag(a  , k=-1)  

array([4, 8])

Create a diagonal matrix from 1D array

In [55]:
v = np.array([10, 20, 30])
np.diag(v)

array([[10,  0,  0],
       [ 0, 20,  0],
       [ 0,  0, 30]])

Replace elements along the main diagonal

In [56]:
a = np.arange(16).reshape(4,4)
np.fill_diagonal(a, 0)
a

array([[ 0,  1,  2,  3],
       [ 4,  0,  6,  7],
       [ 8,  9,  0, 11],
       [12, 13, 14,  0]])

<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/3.png" 
       alt="Sample Image" 
       style="max-width: 60%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


<div style="width: 100px; height: 100px;  ">
</div>


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #8026adff, #cb8dddff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 150px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.logical
    </h2>
  </div>
</div>


 

**np.logical_**
 
NumPy provides several element-wise logical functions to perform boolean operations on arrays.  
Common functions are:  
- `np.logical_and(x, y)` : Element-wise AND  
- `np.logical_or(x, y)`  : Element-wise OR  
- `np.logical_not(x)`    : Element-wise NOT  
- `np.logical_xor(x, y)` : Element-wise XOR

 

 

```python
 

# Boolean arrays
a = np.array([True, False, True, False])
b = np.array([True, True, False, False])

print("AND :", np.logical_and(a, b))  # [ True False False False]
print("OR  :", np.logical_or(a, b))   # [ True  True  True False]
print("NOT a:", np.logical_not(a))    # [False  True False  True]
print("XOR :", np.logical_xor(a, b))  # [False  True  True False]

# Numerical arrays with conditions
x = np.array([1, 2, 3, 4, 5, 6])

# Elements between 2 and 5
mask = np.logical_and(x > 2, x < 5)
print("Between 2 and 5:", x[mask])  # [3 4]

# Elements less than 2 or greater than 5
mask2 = np.logical_or(x < 2, x > 5)
print("Less than 2 or greater than 5:", x[mask2])  # [1 6]

# Using logical_not
print("Not greater than 3:", x[np.logical_not(x > 3)])  # [1 2 3]


&nbsp;

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #720a45ff, #69203cff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 100px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.take
    </h2>
  </div>
</div>


1️⃣ np.take
**What it does:**  
Select specific elements by **index** along an axis. Useful for **reordering data or picking rows/columns**. <br>
انتخاب عناصر خاص با **ایندکس** در امتداد یک محور. برای مرتب‌سازی مجدد داده‌ها یا انتخاب ردیف‌ها / ستون‌ها مفید است

In [57]:
data = np.array([[10, 20, 30],
                 [40, 50, 60],
                 [70, 80, 90]])
reordered = np.take(data, [2,0,7])
print(reordered)

[30 10 80]


In [58]:
selected_rows = np.take(data, [0,2], axis=0)
print(selected_rows)

[[10 20 30]
 [70 80 90]]


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #37ab35ff, #98e09aff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 178px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.compress
    </h2>
  </div>
</div>


2️⃣ np.compress

What it does: <br>
Select elements, rows, or columns where a boolean condition is True. Great for filtering data dynamically

In [81]:
data = np.array([[10, 20, 30],
                 [40, 50, 60],
                 [70, 80, 90]])
mask = data.sum(axis=1) > 100
filtered_rows = np.compress(mask, data, axis=0)
filtered_rows


array([[40, 50, 60],
       [70, 80, 90]])

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #e6de0cff, #dde192ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 245px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.fromfunction
    </h2>
  </div>
</div>


 🔹 `np.fromfunction(function, shape, dtype=float)`

 Syntax:
**np.fromfunction**(function, shape, dtype=float)

Parameters:
- **function**: A lambda or function that receives indices (i, j, k, ...)  
  Example: `lambda i, j: i + j`

- **shape**: Tuple defining the shape of the array  
  Example: `(3, 3)`, `(4, 5)`, `(2, 3, 4)`

- **dtype**: Data type of output (int, float, bool, ...)

---

📝 Common Patterns:

1. **Sum of indices** → `λ i, j: i + j`  
   👉 Creates a matrix where each element = row + col

2. **Multiplication of indices** → `λ i, j: i * j`  
   👉 Multiplication table

3. **Identity matrix** → `λ i, j: i == j`  
   👉 Equivalent to `np.eye(n)`

4. **Mathematical functions** → `λ i, j: np.sin(i) + np.cos(j)`

5. **Higher dimensions** → `λ i, j, k: i + j + k`  
   👉 Works for N-D arrays

Examples:

```python
# Sum of indices (3x3)
arr = np.fromfunction(lambda i, j: i + j, (3, 3))
# [[0 1 2]
#  [1 2 3]
#  [2 3 4]]

# Multiplication table
mult_table = np.fromfunction(lambda i, j: i * j, (5, 5), dtype=int)

# Identity matrix
identity = np.fromfunction(lambda i, j: i == j, (4, 4), dtype=int)

# Mathematical pattern
pattern = np.fromfunction(lambda i, j: np.sin(i) + np.cos(j), (3, 4))
```

---

### ⚡ Tips:
- Indices are floats by default → set `dtype=int` if needed
- Very useful for generating structured test arrays
- Great for quick matrix generation (like grids, eye, etc.)
 

<br>

---

&nbsp;

&nbsp;
&nbsp;
&nbsp;

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #320947ff, #54305fff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 340px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      argmin and argmax 
    </h2>
  </div>
</div>



<div style="text-align: center;"><img src="png/4.png" alt="Local Image" style="max-width:70%; height:auto; border-radius:12px;"><div>

📌 Syntax: <br>
np.argmin(arr, axis=None)   → returns index of minimum<br>
np.argmax(arr, axis=None)   → returns index of maximum

📌 Parameters:<br>
axis=None → کل آرایه<br>
axis=0    → هر ستون<br>
axis=1    → هر ردیف

📌 Example Code:<br>

In [60]:
arr = np.array([10, 5, 20, 3, 15])
print("argmin:", np.argmin(arr))  # 3
print("argmax:", np.argmax(arr))  # 2

matrix = np.array([[1,5,2],
                   [7,0,6]])
print("argmin axis=0:", np.argmin(matrix, axis=0))  # [0 1 0]
print("argmax axis=1:", np.argmax(matrix, axis=1))  # [1 2]

# استفاده برای گرفتن مقدار واقعی
min_value = arr[np.argmin(arr)]
max_value = arr[np.argmax(arr)]

argmin: 3
argmax: 2
argmin axis=0: [0 1 0]
argmax axis=1: [1 0]


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #f56618ff, #feb892ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 100px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      argsort 
    </h2>
  </div>
</div>


&nbsp;
&nbsp;
&nbsp;

📌 Syntax: <br>
**np.argsort**(arr, axis=-1)   → returns indices that would sort the array <br>

📌 Parameters: <br>
axis=None → flatten array <br> 
axis=0    → sort each column <br>
axis=1    → sort each row

In [61]:
arr = np.array([40, 10, 30, 20])
indices = np.argsort(arr)      # اندیس‌هایی که آرایه را صعودی مرتب می‌کنند
print(arr[indices])            # [10 20 30 40]

# مرتب‌سازی نزولی
indices_desc = np.argsort(arr)[::-1]
print(arr[indices_desc])       # [40 30 20 10]

# مرتب کردن آرایه مرتبط با نمرات
scores = np.array([70, 90, 50, 80])
names = np.array(['Ali', 'Sara', 'John', 'Mary'])
sorted_names = names[np.argsort(scores)]
print(sorted_names)            # ['John' 'Ali' 'Mary' 'Sara']

[10 20 30 40]
[40 30 20 10]
['John' 'Ali' 'Mary' 'Sara']


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #1129a3ff, #574b9fff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 190px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.nonzero
    </h2>
  </div>
</div>


1️⃣ np.nonzero(arr) <br>
- Returns indices of non-zero elements. <br>
- Output: tuple of arrays (one per dimension)

In [62]:
a = np.array([0, 2, 0, 5, 3])
indices = np.nonzero(a)
print(indices)
print(a[indices])


(array([1, 3, 4]),)
[2 5 3]


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #570738ff, #583049ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 270px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.count_nonzero
    </h2>
  </div>
</div>


2️⃣ np.count_nonzero(arr, axis=None)
- Counts number of non-zero elements.
- axis=None → all array
- axis=0 or 1 → along column/row (for 2D arrays)

In [63]:
matrix = np.array([[0, 1, 2],
                   [3, 0, 0]])
print(np.count_nonzero(matrix))       
print(np.count_nonzero(matrix, axis=0))   
print(np.count_nonzero(matrix, axis=1)) 

3
[1 1 1]
[2 1]


<a href="https://chatgpt.com/s/t_68aecdaacfb88191a2dc58df43d44877" 
   target="_blank" 
   style="color: #e29cbaff; text-decoration: none; font-size: 40px;">
   np.linalg link
</a>


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #5ddcf3ff, #93e2f0ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 110px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.sort
    </h2>
  </div>
</div>


 NumPy Sort Functions Reference

📌 `np.sort(a, axis=-1, kind='quicksort', order=None)`

- **Returns**: sorted array
- **axis=-1** → last axis (default)
- **kind** → `'quicksort'`, `'mergesort'`, `'heapsort'`, `'stable'`
- **order** → for structured arrays

In [64]:
# Basic sorting
arr = np.array([3,1,4,2])
np.sort(arr)  # [1 2 3 4]

# Matrix sorting
matrix = np.array([[3,1,2],[6,4,5]])
np.sort(matrix, axis=1)  # row-wise: [[1 2 3],[4 5 6]]

# Structured array sorting
data = np.array([(3,'b'),(1,'a'),(2,'c')],
                dtype=[('x',int),('y','U1')])
np.sort(data, order='x')  # [(1,'a'),(2,'c'),(3,'b')]

array([(1, 'a'), (2, 'c'), (3, 'b')], dtype=[('x', '<i8'), ('y', '<U1')])

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #045a11ff, #4a6c4fff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 245px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.sort_complex
    </h2>
  </div>
</div>


📌 `np.sort_complex(a)`

- **Purpose**: Sorts complex numbers by real part, then imaginary

Example:

```python
arr = np.array([3+2j,1+5j,3+1j])
np.sort_complex(arr)  # [1+5j, 3+1j, 3+2j]
```

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #f200b1ff, #eb8ad1ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 500px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.median and np.percentile
    </h2>
  </div>
</div>


 

📌 `np.median(a, axis=None)`

- **Returns**: median of array
- **axis=None** → median of flattened array
- **axis=0** → column-wise median
- **axis=1** → row-wise median

 

```python
a = np.array([1,3,5,7,9])
np.median(a)  # 5

matrix = np.array([[1,3,5],[2,4,6]])
np.median(matrix, axis=0)  # [1.5 3.5 5.5]
np.median(matrix, axis=1)  # [3 4]
```

---

📌 `np.percentile(a, q, axis=None, interpolation='linear')`

- **Returns**: q-th percentile of array
- **q** = 0-100

 

```python
a = np.array([1,2,3,4,5,6,7,8,9,10])
np.percentile(a, 25)  # 25th percentile → 3.25
np.percentile(a, 50)  # 50th percentile → 5.5
np.percentile(a, 75)  # 75th percentile → 7.75
```

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #700027ff, #b60c50ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 140px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.select
    </h2>
  </div>
</div>


`np.select(conditions, choices, default=0)`

- **conditions**: list of boolean arrays
- **choices**: list of values/arrays corresponding to conditions
- **default**: value if no condition is True

In [65]:
import numpy as np

x = np.array([1, 2, 3, 4, 5])

conditions = [x < 3, x == 3, x > 3]
choices = ['small', 'medium', 'large']

result = np.select(conditions, choices, default='unknown')
result

array(['small', 'small', 'medium', 'large', 'large'], dtype='<U7')

In [66]:
scores = np.array([45, 67, 89, 55, 76])
conditions = [scores < 50, (scores >= 50) & (scores < 70), scores >= 70]
grades = [1, 2, 3]  # 1=Fail, 2=Pass, 3=Excellent

result = np.select(conditions, grades, default=0)
result

array([1, 2, 3, 2, 3])

<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/5.png" 
       alt="Sample Image" 
       style="max-width: 100%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #e6720cff, #f1b265ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 180px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.piecewise
    </h2>
  </div>
</div>


 `np.piecewise(x, condlist, funclist)`

- **x**: array to apply piecewise function
- **condlist**: list of boolean arrays or condition functions
- **funclist**: list of functions or values for each condition

In [67]:
x = np.array([0, 1, 2, 3, 4, 5])

conditions = [x < 2, (x >= 2) & (x < 4), x >= 4]
functions = [lambda x: x**2, lambda x: x+10, lambda x: x*2]

y = np.piecewise(x, conditions, functions)
y


array([ 0,  1, 12, 13,  8, 10])

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #9501dfff, #ae4dccff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 100px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.isin
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/6.png" 
       alt="Sample Image" 
       style="max-width: 50%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


In [68]:
a = np.array([[1, 2], [3, 4]])
b = [1, 4]

result = np.isin(a, b)
result


array([[ True, False],
       [False,  True]])

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #1fc7f5ff, #94e3f5ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 140px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.unique
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/7.png" 
       alt="Sample Image" 
       style="max-width: 60%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


In [69]:
a = np.array([1,2,2,3,3,3,4])
np.unique(a)

array([1, 2, 3, 4])

In [70]:
unique, counts = np.unique(a, return_counts=True)
print(unique, counts)

[1 2 3 4] [1 2 3 1]


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #89f557ff, #e9f57aff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 140px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.repeat
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/8.png" 
       alt="Sample Image" 
       style="max-width: 60%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


In [71]:
a = np.array([1,2,3])
np.repeat(a, 2)

array([1, 1, 2, 2, 3, 3])

In [72]:
b = np.array([[1,2],[3,4]])
np.repeat(b, 2, axis=0)  # تکرار ردیف‌ها

array([[1, 2],
       [1, 2],
       [3, 4],
       [3, 4]])

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #320947ff, #54305fff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 100px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.clip
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/9.png" 
       alt="Sample Image" 
       style="max-width: 50%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


In [73]:
a = np.array([1, 5, 10, 15, 20])
np.clip(a, 5, 15)

array([ 5,  5, 10, 15, 15])

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #895b6aff, #9673a1ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 150px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.round
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/10.png" 
       alt="Sample Image" 
       style="max-width: 50%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


In [74]:
b = np.array([1.234, 2.567, 3.891])
np.round(b, 2)

array([1.23, 2.57, 3.89])

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #769bc0ff, #d7adceff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 200px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.corrcoef
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/11.png" 
       alt="Sample Image" 
       style="max-width: 50%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


In [75]:
import numpy as np

x = [1, 2, 3, 4, 5]
y = [2, 4, 6, 8, 10]
corr = np.corrcoef(x, y)
corr

array([[1., 1.],
       [1., 1.]])

<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #6d7cccff, #b559d1ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 100px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.cov
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/12.png" 
       alt="Sample Image" 
       style="max-width: 50%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


In [76]:
x = [1, 2, 3, 4, 5]
y = [2, 4, 6, 8, 10]
cov = np.cov(x, y)
cov

array([[ 2.5,  5. ],
       [ 5. , 10. ]])

 <div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #6d7cccff, #272d45ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 220px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.concatenate
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/13.png" 
       alt="Sample Image" 
       style="max-width: 50%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #bbc0daff, #876d8fff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 100px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.split
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/14.png" 
       alt="Sample Image" 
       style="max-width: 50%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #021ec0ff, #520c68ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 140px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.vstack
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/15.png" 
       alt="Sample Image" 
       style="max-width: 50%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #1f1b9aff, #551f65ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 140px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.hstack
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/16.png" 
       alt="Sample Image" 
       style="max-width: 50%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #315514ff, #cfc8a3ff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 125px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.stack
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/17.png" 
       alt="Sample Image" 
       style="max-width: 50%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


<div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #041261ff, #c5ccedff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 100px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      np.put
    </h2>
  </div>
</div>


<div style="display: flex; justify-content: center; margin: 20px 0;">
  <img src="png/18.png" 
       alt="Sample Image" 
       style="max-width: 50%; height: auto; border-radius: 12px; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
</div>


In [77]:
arr = np.array([10, 20, 30, 40, 50])
np.put(arr, [1, 3], [200, 400])  # تغییر ایندکس‌های 1 و 3
arr

array([ 10, 200,  30, 400,  50])

 <div style="display: flex; justify-content: center; width: 100%;">
  <div style="background: linear-gradient(135deg, #55053aff, #22033dff); padding: 25px 40px; border-radius: 12px; text-align: center; color: white; max-width: 700px; width: 100%; box-shadow: 0 4px 10px rgba(0,0,0,0.2);">
    <h2 style="margin: 0; font-size: 2em; font-weight: 600; text-shadow: 1px 1px 3px rgba(0,0,0,0.4);">
      NumPy mathematical and aggregation operations
    </h2>
  </div>
</div>


```python 
import numpy as np

a = np.array([1, 2, 3, 4, 5])
b = np.array([5, 4, 3, 2, 1])

# 1️⃣ np.add
# English: Adds two arrays element-wise
add_result = np.add(a, b)  # [6 6 6 6 6]

# 2️⃣ np.multiply
# English: Multiplies two arrays element-wise
mul_result = np.multiply(a, b)  # [5 8 9 8 5]

# 3️⃣ np.divide
# English: Divides two arrays element-wise
div_result = np.divide(a, b)  # [0.2 0.5 1. 2. 5.]

# 4️⃣ np.power
# English: Raises elements of first array to powers from second array
pow_result = np.power(a, b)  # [1 16 27 16 5]

# 5️⃣ np.mod
# English: Element-wise modulo
mod_result = np.mod(a, b)  # [1 2 0 0 0]

# 6️⃣ np.cumsum
# English: Cumulative sum
cumsum_result = np.cumsum(a)  # [1 3 6 10 15]

# 7️⃣ np.cumprod
# English: Cumulative product
cumprod_result = np.cumprod(a)  # [1 2 6 24 120]

# 8️⃣ np.diff
# English: Calculates the difference between consecutive elements
diff_result = np.diff(a)  # [1 1 1 1]
 

# 🔟 np.round / np.floor / np.ceil
# English: Rounds numbers to nearest integer, floor or ceil
arr = np.array([1.2, 2.5, 3.7])

round_result = np.round(arr)  # [1. 2. 4.]  
floor_result = np.floor(arr)  # [1. 2. 3.]   
ceil_result  = np.ceil(arr)   # [2. 3. 4.]   
 